src/Everscale/CellBuilder.cpp

```diff
@@ -0,0 +1,260 @@
+            throw std::runtime_error("cell refs overflow");
+        }
+        references.emplace_back(std::move(child));
+    }
+}
+
+void CellBuilder::appendBuilder(const CellBuilder& builder) {
+    appendRaw(builder.data, builder.bitLen);
+    for (const auto& reference : builder.references) {
+        appendReferenceCell(reference);
+    }
+}
+
+void CellBuilder::appendCellSlice(const CellSlice& other) {
+    Data appendedData(other.cell->data);
+    appendRaw(appendedData, other.cell->bitLen);
+
+    for (const auto& cell : other.cell->references) {
+        appendReferenceCell(cell);
+    }
+}
+
+Cell::Ref CellBuilder::intoCell() {
+    // Append tag
+    if (bitLen & 7) {
+        const auto mask = static_cast<uint8_t>(0x80 >> (bitLen & 7));
+        const auto l = bitLen / 8;
+        data[l] = static_cast<uint8_t>((data[l] & ~mask) | mask);
+    }
+
+    auto refCount = references.size();
+
+    Cell::Refs refs;
+    std::move(references.begin(), references.end(), refs.begin());
+
+    auto cell = std::make_shared<Cell>(bitLen, std::move(data), refCount, std::move(refs));
+    cell->finalize();
+
+    bitLen = 0;
+    refCount = 0;
+
+    return cell;
+}
+
+void CellBuilder::appendWithoutShifting(const Data& appendedData, uint16_t bits) {
+    assert(bits % 8 == 0);
+    assert(bitLen % 8 == 0);
+
+    data.resize(bitLen / 8);
+    data.insert(data.end(), appendedData.begin(), appendedData.end());
+    bitLen += bits;
+    data.resize(bitLen / 8);
+}
+
+void CellBuilder::appendWithSliceShifting(const Data& appendedData, uint16_t bits) {
+    assert(bits % 8 != 0);
+    assert(bitLen % 8 == 0);
+
+    data.resize(bitLen / 8);
+    data.insert(data.end(), appendedData.begin(), appendedData.end());
+    bitLen += bits;
+    data.resize(1 + bitLen / 8);
+
+    data.back() &= ~static_cast<uint8_t>(0xff >> (bits % 8));
+}
+
+void CellBuilder::appendWithDoubleShifting(const Data& appendedData, uint16_t bits) {
+    auto selfShift = bitLen % 8;
+    data.resize(1 + bitLen / 8);
+    bitLen += bits;
+
+    // yyyyy000 -> 00000000 000yyyyy
+    auto y = static_cast<uint16_t>(data.back() >> (8 - selfShift));
+    data.pop_back();
+
+    for (const auto x : appendedData) {
+        // 00000000 000yyyyy -> 000yyyyy xxxxxxxx
+        y = static_cast<uint16_t>(y << 8) | x;
+        // 000yyyyy xxxxxxxx -> 00000000 yyyyyxxx
+        data.push_back(static_cast<uint8_t>(y >> selfShift));
+    }
+    // 00000000 yyyyyxxx
+    data.push_back(static_cast<uint8_t>(y << (8 - selfShift)));
+
+    auto shift = bitLen % 8;
+    if (shift == 0) {
+        data.resize(bitLen / 8);
+    } else {
+        data.resize(bitLen / 8 + 1);
+        data.back() &= ~static_cast<uint8_t>(0xff >> (bitLen % 8));
+    }
+}
+
+uint8_t CellBuilder::clzU128(const uint128_t& u) {
+    auto hi = static_cast<uint64_t>(u >> 64);
+    auto lo = static_cast<uint64_t>(u);
+
+    int retval[3] = {
+        __builtin_clzll(hi),
+        __builtin_clzll(lo) + 64,
```

Comment:
Could you please use https://en.cppreference.com/w/cpp/numeric/countl_zero here instead?

Reply:
Done in https://github.com/trustwallet/wallet-core/pull/2428/commits/423763fbca3e97356083d7bec1c543cbfa9bd906